Add namespace-name suggestions to C# and VB

lorcanmooney opened the pull request:
Part of #7213, this adds namespace-name suggestions to C# and VB.

lorcanmooney commented on src/Workspaces/VisualBasic/Portable/Recommendations/VisualBasicRecommendationService.vb:
```diff
@@ -212,6 +212,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Recommendations
             Else
                 symbols = context.SemanticModel _
```
I don't fully understand this `couldBeMergedNamespace` thing or it's consequences, but I couldn't hit the positive code path. Am I missing something I'm missing here?

rchande replied:
The `couldBeMergedNamespace` thing is an error recovery scenario where `GetSymbolInfo` returns a bunch of `CandidateSymbols` that are all namespaces. I tried chasing through source control to find the original related bug but didn't find it, but I don't think you need to worry about it.

lorcanmooney commented:
Tagging @CyrusNajmabadi, @rchande,  @DustinCampbell and @Pilchie as potential reviewers.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -603,6 +603,18 @@ public static bool IsAttributeNameContext(this SyntaxTree syntaxTree, int positi
 
+        public static bool IsNamespaceDeclarationNameContext(this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        {
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
```
i think you need a ".GetPreviousTokenIfTouchingWord(position)".  That way if you're actually *on* the 'namespace' keyword itself, that we don't think we're in the namespcae name portion.

lorcanmooney replied:
Huh, you're right about it returning the wrong value when _on_ `namespace`, but it doesn't seem to be causing any problems. Subsequent checks in the recommendation service must be mitigating it. I'll need to investigate this a bit more.

lorcanmooney replied:
Oh, good catch! It doesn't show up in the simple case, but if we have nested namespaces...
```
namespace Microsoft
{
	namespace Code$$Analysis
	{
	}
}
```
...then it starts suggesting top level namespaces! I think this deserves a few additional tests.

lorcanmooney replied:
Fixed. I've also added a check to bail early on scripts, since apparently they don't support namespace declaration.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationService.cs:
```diff
@@ -321,6 +323,12 @@ internal class CSharpRecommendationService : AbstractRecommendationService
 
+                if (context.IsNamespaceDeclarationNameContext)
+                {
+                    NamespaceDeclarationSyntax declarationSyntax = name.GetAncestorOrThis<NamespaceDeclarationSyntax>();
```
var.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationService.cs:
```diff
@@ -238,6 +234,12 @@ internal class CSharpRecommendationService : AbstractRecommendationService
+        private static IEnumerable<ISymbol> GetSymbolsForNamespaceDeclarationNameContext(CSharpSyntaxContext context, CancellationToken cancellationToken)
+        {
+            var declarationSyntax = context.TargetToken.GetAncestor<NamespaceDeclarationSyntax>();
+            return GetRecommendedNamespaceNameSymbols(context.SemanticModel, declarationSyntax, cancellationToken);
```
What if there is no containing namespace declaration?

lorcanmooney replied:
It should only be called when we're sure we're already in an appropriate context. I'll add a check to harden in against misuse.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationService.cs:
```diff
@@ -29,6 +29,26 @@ internal abstract class AbstractRecommendationService : IRecommendationService
+
+        protected static bool IsNonIntersectingNamespace(ISymbol recommendationSymbol, SyntaxNode declarationSyntax)
+        {
+            return recommendationSymbol.IsNamespace() &&
```
Why do we need the IsNonIntersectingNamespace check?  Can you comment teh code as to why we need this?  THanks!

lorcanmooney replied:
It just filters out the declaration the user is in the middle of typing, and for VB it filters out any namespaces nested within an existing declaration (since they can never be peers of the declaration).

I'll add the comment.

CyrusNajmabadi replied:
Could you give an example.  I'm not quite sure i understand hte case you're talking about.  Thanks!

lorcanmooney replied:
```
Namespace Foo$$
End Namespace
```
In the above `Foo` shouldn't be considered for the completion list unless it's also declared somewhere else. This applies to VB and C#.

```
Namespace Foo.$$
    Namespace Bar
    End Namespace
End Namespace
```
Here, `Bar` shouldn't be included the completion, because it will always be a child of whatever the user is about to type (again, unless it's also declared somewhere else). For whatever reason this only applies to how VB presents the symbols. Regardless, I added tests for both languages.

CyrusNajmabadi replied:
> Here, Bar shouldn't be included the completion, 

I'm not sure about htat.  I may be deciding that i want to merge 'Bar' into the 'Foo' declaration.  So i'm going to complete out to 'Foo.Bar', then delete the namespace statement for ```namespace Bar```.

I'm not certain this filtering is necessary at all.

lorcanmooney replied:
How about this?
```
Namespace Foo.$$
    Namespace Bar
    End Namespace

    Namespace Baz
    End Namespace
End Namespace
```
Should `Baz` also show up in the list?

It's also worth noting that getting equivalent behaviour in C# will probably need additional lookups.

CyrusNajmabadi replied:
Sure.  I wouldn't fret at all about these items showing up in the completion list.

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.vb:
```diff
@@ -61,6 +61,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions.ContextQuery
         <Extension()>
+        Public Function IsNamespaceDeclarationNameContext(syntaxTree As SyntaxTree, position As Integer, cancellationToken As CancellationToken) As Boolean
+            Dim statement = syntaxTree.GetTargetToken(position, cancellationToken).GetAncestor(Of NamespaceStatementSyntax)
+            Return statement IsNot Nothing AndAlso statement.Name.Span.IntersectsWith(position)
```
Shouldn't this check if the previous token is the namespcae keyword as well?

lorcanmooney replied:
I don't think so. Is the `Name.Span` check not be sufficient?

CyrusNajmabadi replied:
1. I'm not thrilled that this code differs from teh C# version. 
2. I'm not sure how sufficient it would be when the user just types ```Namespace``` and doesn't have a name.  Now you're going to be checking the span of a missing node.  And those are notoriously finicky.

I'm happy to use the span if the name is there.  But if it isn't, we should explicitly make sure the code works for the case it isn't.

lorcanmooney replied:
1. Assuming it works, I'll make it identical to the C# logic (i.e. use `FindTokenOnLeftOfPosition`) but...
2. ...the C# code also uses the name span. That's not actually my addition; it's been in the product since the initial release, used to _disable_ the completion list.

The tests cover the cases of having no name and incomplete declarations for both languages - and I'll add a few more variations for good measure - so any future change to the behaviour should be caught.

lorcanmooney replied:
Forget everything I said there, after fixing the C# version, the empty span _did_ start causing trouble. The VB version is now identical to the fixed C# method.

CyrusNajmabadi commented:
Overall looking good!  I had a few small questions/concerns.  Thanks.

lorcanmooney commented:
@CyrusNajmabadi Thanks, I'll fix those up and add comments as appropriate. Do you have any suggestions regarding my question about the `couldBeMergedNamespace` path?

lorcanmooney commented:
@CyrusNajmabadi  I've updated the PR based your feedback. I've fixed everything except the `IsNonIntersectingNamespace` method, since (a) it will re-introduce a problem I was having with incorrect symbols a while ago (see #7213), (b) I'm not convinced the effort is worth it, and (c) it's late here and I need some sleep! I hope it won't block the merge.

CyrusNajmabadi commented:
test vsi please

CyrusNajmabadi commented:
:+1: 

Nice job!

Pilchie commented:
@rchande and @dotnet/roslyn-ide can you take a look?

rchande commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SnippetCompletionProvider.cs:
```diff
@@ -119,16 +119,17 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
         }
 
-        private async Task<IEnumerable<CompletionItem>> GetSnippetCompletionItemsAsync(Workspace workspace, SemanticModel semanticModel, int position, TextSpan itemSpan, bool isPreProcessorContext, CancellationToken cancellationToken)
+        private async Task<IEnumerable<CompletionItem>> GetSnippetCompletionItemsAsync(Workspace workspace, SemanticModel semanticModel, TextSpan itemSpan, bool isPreProcessorContext, CancellationToken cancellationToken)
```
Unrelated?

lorcanmooney replied:
Yes, unrelated. I noticed the param was unused as I was passing through. Should I revert it?

CyrusNajmabadi replied:
Don't need to revert this.

rchande replied:
No, it's fine.

rchande commented:
:+1:

rchande commented:
@dotnet-bot retest windows_vsi_p1_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p1_open_prtest/120/
// Retest reason:

lorcanmooney commented:
I don't have access to the vsi_p2 test failure results, so please let me know if there's something I need to investigate.

CyrusNajmabadi commented:
Nope.  It's a known issue.

CyrusNajmabadi commented:
retest windows_vsi_p2_open_prtest  please

CyrusNajmabadi commented:
retest windows_vsi_p2_open_prtest  please

CyrusNajmabadi commented:
Thanks Lorcan!
